File: companies/ByteDance/210.course-schedule-ii.go
package leetcode

// FindOrder210 is Kahn's topological sort and returns the processed order. If a
// cycle prevents processing all courses, the valid order does not exist.
//
// Time: O(V+E). Space: O(V+E).
func FindOrder210(numCourses int, prerequisites [][]int) []int {
	graph := make([][]int, numCourses)
	indeg := make([]int, numCourses)
	for _, p := range prerequisites {
		course, pre := p[0], p[1]
		graph[pre] = append(graph[pre], course)
		indeg[course]++
	}
	q := []int{}
	for i, d := range indeg {
		if d == 0 {
			q = append(q, i)
		}
	}
	order := []int{}
	for head := 0; head < len(q); head++ {
		node := q[head]
		order = append(order, node)
		for _, nxt := range graph[node] {
			indeg[nxt]--
			if indeg[nxt] == 0 {
				q = append(q, nxt)
			}
		}
	}
	if len(order) != numCourses {
		return []int{}
	}
	return order
}

File: companies/ByteDance/210.course-schedule-ii.java
import java.util.*;

/**
 * Algorithm:
 * Topological sort with indegrees. Append each zero-indegree course to the
 * order and remove its outgoing edges. If a cycle remains, not all courses are
 * output and the answer is empty.
 *
 * Java data structures:
 * ArrayList<Integer>[] stores adjacency; ArrayDeque<Integer> stores the current
 * zero-indegree frontier.
 *
 * Complexity:
 * Time O(V + E), space O(V + E).
 */
class Solution {
    public int[] findOrder(int numCourses, int[][] prerequisites) {
        List<Integer>[] graph = new ArrayList[numCourses];
        for (int i = 0; i < numCourses; i++) {
            graph[i] = new ArrayList<>();
        }
        int[] indegree = new int[numCourses];
        for (int[] edge : prerequisites) {
            int course = edge[0];
            int pre = edge[1];
            graph[pre].add(course);
            indegree[course]++;
        }
        Queue<Integer> q = new ArrayDeque<>();
        for (int i = 0; i < numCourses; i++) {
            if (indegree[i] == 0) {
                q.offer(i);
            }
        }
        int[] order = new int[numCourses];
        int size = 0;
        while (!q.isEmpty()) {
            int node = q.poll();
            order[size++] = node;
            for (int next : graph[node]) {
                if (--indegree[next] == 0) {
                    q.offer(next);
                }
            }
        }
        return size == numCourses ? order : new int[0];
    }
}

